File: 2025/1438. Longest Continuous Subarray With Absolute Diff Less Than or Equal to Limit/Solution.java
class Solution {
    public int longestSubarray(int[] nums, int limit) {
        Deque<Integer> increaseQueue = new ArrayDeque<>(), decreaseQueue = new ArrayDeque<>();
        int res = 0, l = 0;
        for (int r = 0; r < nums.length; r++) {
            int val = nums[r];
            while (!increaseQueue.isEmpty() && increaseQueue.peekLast() > val) {
                increaseQueue.pollLast();
            }
            while (!decreaseQueue.isEmpty() && decreaseQueue.peekLast() < val) {
                decreaseQueue.pollLast();
            }
            increaseQueue.offerLast(val);
            decreaseQueue.offerLast(val);
            if (Math.abs(increaseQueue.peekFirst() - decreaseQueue.peekFirst()) <= limit) {
                res = Math.max(res, r - l + 1);
            } else {
                if (increaseQueue.peekFirst() == nums[l]) {
                    increaseQueue.pollFirst();
                }
                if (decreaseQueue.peekFirst() == nums[l]) {
                    decreaseQueue.pollFirst();
                }
                l++;
            }
        }
        return res;
    }
}

File: 2025/1438. Longest Continuous Subarray With Absolute Diff Less Than or Equal to Limit/solution.py
class Solution:
    def longestSubarray(self, nums: List[int], limit: int) -> int:
        # monotonoic increasing and decreasing queues
        min_queue, max_queue = deque(), deque()
        l = r = longest = 0
        while r < len(nums):
            val = nums[r]
            while max_queue and max_queue[-1] < val:
                max_queue.pop()
            while min_queue and min_queue[-1] > val:
                min_queue.pop()
            max_queue.append(val)
            min_queue.append(val)
            if (abs(max_queue[0] - min_queue[0]) <= limit):
                longest = max(longest, r - l + 1)
            else:
                if max_queue and max_queue[0] == nums[l]:
                    max_queue.popleft()
                if min_queue and min_queue[0] == nums[l]:
                    min_queue.popleft()
                l += 1
            r += 1
        return longest